mautic/mautic
# [Enhancement] Allow filtering contacts by Campaign Membership for segments

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | Yes
| Automated tests included? | No
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | #5909
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
Continuing from #5868, #5886 we sometimes need to cut a new segment (or make a report) based on the campaign/s of certain leads after ingestion.

For example, we may want to create a campaign that directly affects leads that were pumped through campaigns A & B without worrying about how the leads got there (through any segment, form, import, etc).

This should work exactly like Segment Membership, and exclude leads that are manually removed from a campaign. Once a lead is "in" a campaign, they should always have that membership till manually removed (to stay forward compatible with rotations).

#### Steps to test this PR:
0. Run `php console mautic:assets:generate`
1. Create a new Segment called "Empty" with a filter "Date Identified" "empty". Users will never end up in this segment unless manually added.
2. Create a new Campaign called "Test Campaign" that takes leads from the "Empty" Segment, and does any action (such as add 1 point).
3. Create another Segment called "Has been in Test Campaign". Use the field "Campaign Membership" and choose our "Test Campaign". This segment should inherit all contacts that are ingested or manually added to the campaign in question.
3. Go to the Contacts list, and click the checkbox next to any contact name, then select from the dropdown above to add this contact to a campaign. Click "Change Campaigns" to add the contact to the campaign "Test Campaign".
4. Run typical `php console mautic:segments:update` or wait till your cron jobs do the same.
5. Go to Segments, and look at "Has been in Test Campaign". It should now have one new contact.
6. For bonus points, manually remove the contact from the campaign, and the segment will drop this contact on the next segment update.
